### PR TITLE
sam0/stm32: fix possible uart_nonblocking deadlock

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

If trying to write to tsrb while ISR are masked or in an ISR then the while loop would block. Avoid blocking if any of those conditions are filled, this might mean bytes are lost.

### Testing procedure

`irq_disable()` in an application printing some output.  I tested in examples hello-world:

```
#include <stdio.h>
#include "irq.h"

int main(void)
{
    unsigned state = irq_disable();
    puts("Hello World!");
    irq_restore(state);
    printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
    printf("This board features a(n) %s MCU.\n", RIOT_MCU);

    return 0;
}
```

```
2020-06-14 20:35:14,805 # main(): This is RIOT! (Version: 2020.07-devel-1126-g5638e-pr_uart_nb_race)
2020-06-14 20:35:14,809 # You are running RIOT on a(n) samr21-xpro board.
2020-06-14 20:35:14,812 # This board features a(n) samd21 MCU.
```

In this example I didn't expect the output to be lost though. Not sure if this is the best solution either.

### CODE REVIEW

**Comment by fjmolinas** [comment]
Maybe in those cases it should flush the necessary amount of bytes?

**Comment by kaspar030** [comment]
> If trying to write to tsrb while ISR are masked or in an ISR then the while loop would block. Avoid blocking if any of those conditions are filled, this might mean bytes are lost.

IMO it needs to fall back to synchronous writes. DMA does that too.

**Comment by kaspar030** [comment]
> Maybe in those cases it should flush the necessary amount of bytes?

ah, that's what you mean!

**Comment by kaspar030** [comment]
> ah, that's what you mean!

ah, so there's two options (plus dropping bytes):

1. flush everything, then write synchronous
2. directly write synchronous

The former would make the ISR take extra time depending on the buffer state, but keep output order.
The latter would make the ISR output show up in between non-ISR output.

I think mixing/re-ordering output would be OK-ish (if surprising) in the stdio use case, but not acceptable when uart is used e.g., for transmitting other data, like network packets.

**Comment by fjmolinas** [comment]
> > ah, that's what you mean!
> 
> ah, so there's two options (plus dropping bytes):
> 
>     1. flush everything, then write synchronous
> 
>     2. directly write synchronous
> 
> 
> The former would make the ISR take extra time depending on the buffer state, but keep output order.
> The latter would make the ISR output show up in between non-ISR output.
> 
> I think mixing/re-ordering output would be OK-ish (if surprising) in the stdio use case, but not acceptable when uart is used e.g., for transmitting other data, like network packets.

I would say there is another:

- flush synchronously the min amount of bytes required in tsrb. (i.e. if we need to write 5 bytes and the ring buffer is full then flush the first 5 bytes in the ring buffer and add the 5 new ones to the ringbuffer)

**Comment by fjmolinas** [comment]
@kaspar030 Did that, know it frees up the required space in the ringbuffer by directly writing the data.

File: cpu/sam0_common/periph/uart.c [hunk 1]
```diff
@@ -160,9 +160,22 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
     return UART_OK;
 }
 
+#ifdef MODULE_PERIPH_UART_NONBLOCKING
+void _flush_uart_tsrb(uart_t uart, size_t len) {
```
Comment (benpicco) [inline cpu/sam0_common/periph/uart.c:164]: `len` is unused

File: cpu/sam0_common/periph/uart.c [hunk 2]
```diff
@@ -160,9 +160,22 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+    int byte = tsrb_get_one(&uart_tx_rb[uart]);
+    dev(uart)->DATA.reg = *byte;
```
Comment (maribu) [inline cpu/sam0_common/periph/uart.c:166]: Should there be some waiting for the byte to actually be send? This seems to be done using the following in the rest of the implementation.

```C
while (!dev(uart)->INTFLAG.bit.DRE) {}
```

**Comment by fjmolinas** [comment]
I pushed somethign that doesn't work sorry.

**Comment by maribu** [comment]
> I pushed somethign that doesn't work sorry.

Happens to me all the time ;-) No need to worry. That what the reviews are for :-)

**Comment by fjmolinas** [comment]
Pusehed some changes;

- 32a0c1a struggled for a while way a test scenario was not working on stm32, and it was because I was not enabling ISR in cases where there was not `rx_cb` set
- ce79ff0 added a test for printing with disabled irq on a full transmit buffer
- 5fee6d3 renamed for convenience `%_UART_TXBUF_SIZE` to `UART_TXBUF_SIZE`.
- I ended up checking if the buffer was full on every byte write attempt because the amount of bytes to write could be bigger than the transmit buffer itself.

<details><summary>BOARD=iotlab-m3 make -C tests/periph_uart_nonblocking/ flash -j3 test
</summary>

```
START
main(): This is RIOT! (Version: 2020.07-devel-1134-gce79f-pr_uart_nb_race)

puts with disabled interrupts and a full transmit buffer

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

Hoarders can get piles of money,
That is true, hackers, that is true.
But they cannot help their neighbors;
That's not good, hackers, that's not good.

When we have enough free software
At our call, hackers, at our call,
We'll kick out those dirty licenses
Ever more, hackers, ever more.

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

== printed in 2101725/2100000 µs ==
[SUCCESS]
```
</details>

<details><summary>BOARD=samr21-xpro make -C tests/periph_uart_nonblocking/ flash -j3 test
</summary>

```
READY
s
START
main(): This is RIOT! (Version: 2020.07-devel-1134-gce79f-pr_uart_nb_race)

puts with disabled interrupts and a full transmit buffer

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

Hoarders can get piles of money,
That is true, hackers, that is true.
But they cannot help their neighbors;
That's not good, hackers, that's not good.

When we have enough free software
At our call, hackers, at our call,
We'll kick out those dirty licenses
Ever more, hackers, ever more.

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

== printed in 2103186/2100000 µs ==
[SUCCESS]
```
</details>

**Comment by maribu** [comment]
Did btw already someone point out that the tread-safe ring buffer is not thread-safe? That might be something to fix rather soonish.

File: cpu/sam0_common/periph/uart.c [hunk 3]
```diff
@@ -164,7 +164,17 @@ void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
 #ifdef MODULE_PERIPH_UART_NONBLOCKING
     for (const void* end = data + len; data != end; ++data) {
-        while (tsrb_add_one(&uart_tx_rb[uart], *data) < 0) {}
+        if (irq_is_in() || __get_PRIMASK()) {
+            /* if ring buffer is full free up a spot */
+            if (tsrb_full(&uart_tx_rb[uart])) {
+                while (!dev(uart)->INTFLAG.bit.DRE) {}
+                dev(uart)->DATA.reg = tsrb_get_one(&uart_tx_rb[uart]);
+                tsrb_add_one(&uart_tx_rb[uart], *data);
```
Comment (benpicco) [inline cpu/sam0_common/periph/uart.c:172]: You want to pull that one out of the `if (tsrb_full())` block or else you are dropping bytes if the buffer is not full.

Comment (fjmolinas) [reply cpu/sam0_common/periph/uart.c]: thanks!

**Comment by maribu** [comment]
> Did btw already someone point out that the tread-safe ring buffer is not thread-safe?

Yes, over a year ago: https://github.com/RIOT-OS/RIOT/issues/9882

File: tests/periph_uart_nonblocking/main.c [hunk 4]
```diff
@@ -30,8 +30,22 @@ static inline uint32_t puts_delay(const char* str)
     return LINE_DELAY_MS * 1000;
 }
 
+static void _irq_disabled_print(void)
+{
+    unsigned state = irq_disable();
+    /* fill the transmit buffer */
+    for(uint8_t i=0; i < UART_TXBUF_SIZE; i ++) {
```
Comment (aabadie) [inline tests/periph_uart_nonblocking/main.c:37]:
```suggestion
    for (uint8_t i = 0; i < UART_TXBUF_SIZE; i++) {
```

File: tests/periph_uart_nonblocking/main.c [hunk 5]
```diff
@@ -30,8 +30,22 @@ static inline uint32_t puts_delay(const char* str)
...
+        printf(" ");
+    }
+    printf("\n");
+    puts("puts with disabled interrupts and a full transmit buffer");
```
Comment (aabadie) [inline tests/periph_uart_nonblocking/main.c:41]:
```suggestion
    puts("\nputs with disabled interrupts and a full transmit buffer");
```

**Comment by aabadie** [comment]
Tested on samr21-xpro, iotlab-m3 and b-l475e-iot01a all worked like a charm.

**Comment by aabadie** [comment]
gentle ping @fjmolinas :)

**Comment by fjmolinas** [comment]
Rebased, reordered the commits a bit.

**Comment by fjmolinas** [comment]
all green @aabadie!

**Comment by aabadie** [comment]
GO!

**Comment by fjmolinas** [comment]
Thanks for the review!

**Comment by maribu** [comment]
I have one more question regarding the use of `tsrb` for UART: The documentation of `tsrb` states:

> This ringbuffer implementation can be used without locking if there's only one producer and one consumer.

The part with one consumer is true: It is the IRQ handler and nested IRQs are not enabled by RIOT (unless someone does so in implementation code).

But e.g. when UART is used for `stdio` and multiple threads are using `stdio`? Am I overlooking something, or does this violate the preconditions for safe use of `tsrb`?

**Comment by kaspar030** [comment]
> But e.g. when UART is used for `stdio` and multiple threads are using `stdio`? Am I overlooking something, or does this violate the preconditions for safe use of `tsrb`?

I think you're right. multiple producers should break the lock-free assumptions. e.g.,

```
static uint8_t _pop(tsrb_t *rb)
{
    return rb->buf[rb->reads++ & (rb->size - 1)];
}


int tsrb_get_one(tsrb_t *rb)
{
    if (!tsrb_empty(rb)) {
        return _pop(rb);
    }
    else {
        return -1;
    }
}
```

with two parallel consumers, both might get tsrb_empty()==false, then both read, even though there's only one byte available.

**Comment by kaspar030** [comment]
> I think you're right. multiple producers should break the lock-free assumptions.

in the stdio use case, this has been in use since 2015 I think. we might have been lucky, as there's usually just the shell as single consumer.

**Comment by maribu** [comment]
> we might have been lucky, as there's usually just the shell as single consumer.

But at least the multiple producer case is pretty common, at least when logging is enabled.

```C
static void _push(tsrb_t *rb, uint8_t c)                                         
{                                                                                
    rb->buf[rb->writes++ & (rb->size - 1)] = c;                                  
}   
```

This is bound to break if multiple producers are there and competing for UART access. (Which is totally fine, as this is clearly documented in TSRB.)

Likely people won't complain in the stdio use case: If the stdio output of multiple threads is being mixed up anyway, it will be hard to even notice a single character that was dropped. I suppose it could become an issue if one would just send single bytes from multiple threads for tracing; a lost byte there could be highly confusing.

**Pull request merged**